cpp/oneapi/dal/algo/pca/backend/gpu/train_kernel_cov_impl_dpc.cpp

```diff
@@ -190,34 +237,59 @@ result_t train_kernel_cov_impl<Float>::operator()(const descriptor_t& desc, cons
 
     if (desc.get_result_options().test(result_options::means)) {
         auto [means, means_event] = compute_means(q_, rows_count_global, sums, { gemm_event });
-        result.set_means(homogen_table::wrap(means.flatten(q_), 1, column_count));
+        result.set_means(homogen_table::wrap(means.flatten(q_, { means_event }), 1, column_count));
     }
 
     auto [cov, cov_event] = compute_covariance(q_, rows_count_global, xtx, sums, { gemm_event });
+
+    auto [vars, vars_event] = compute_variances(q_, cov, { cov_event });
     if (desc.get_result_options().test(result_options::vars)) {
-        auto [vars, vars_event] = compute_variances(q_, cov, { cov_event });
-        vars_event.wait_and_throw();
-        result.set_variances(homogen_table::wrap(vars.flatten(q_), 1, column_count));
+        result.set_variances(
+            homogen_table::wrap(vars.flatten(q_, { vars_event }), 1, column_count));
+    }
+    auto data_to_compute = cov;
+    sycl::event corr_event;
+    if (desc.get_normalization_mode() == normalization::zscore) {
+        auto corr = pr::ndarray<Float, 2>::empty(q_, { column_count, column_count }, alloc::device);
+        std::tie(corr, corr_event) =
+            compute_correlation_from_covariance(q_, rows_count_global, cov, { cov_event });
+        corr_event.wait_and_throw();
+        data_to_compute = corr;
+    }
+
+    auto [eigvecs, eigvals] = compute_eigenvectors_on_host(q_,
+                                                           std::move(data_to_compute),
+                                                           component_count,
+                                                           { cov_event, corr_event, vars_event });
+    if (desc.get_result_options().test(result_options::eigenvalues)) {
+        result.set_eigenvalues(homogen_table::wrap(eigvals.flatten(), 1, component_count));
+    }
+    if (desc.get_result_options().test(result_options::singular_values)) {
+        auto singular_values =
+            compute_singular_values_on_host(q_,
+                                            eigvals,
+                                            row_count,
+                                            { cov_event, corr_event, vars_event });
+        result.set_singular_values(
+            homogen_table::wrap(singular_values.flatten(), 1, component_count));
+    }
+    if (desc.get_result_options().test(result_options::explained_variances_ratio)) {
+        auto vars_host = vars.to_host(q_);
+        auto explained_variances_ratio =
+            compute_explained_variances_on_host(q_,
+                                                eigvals,
+                                                vars_host,
+                                                { cov_event, corr_event, vars_event });
+        result.set_explained_variances_ratio(
+            homogen_table::wrap(explained_variances_ratio.flatten(), 1, component_count));
+    }
```

Comment:
Output from such functions can be unified

Reply:
it will be unified in the separate pr with removing code duplications for cov/cor/means computation